presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryExecutor.java

```diff
@@ -65,6 +86,26 @@ public StatementClient startQuery(ClientSession session, String query)
     @Override
     public void close()
     {
+        if (closer.isPresent()) {
+            closer.get().accept(this);
+        }
+        else {
+            closeHttpClient();
+        }
+    }
+
+    int reference()
```

Comment:
I would rather move the reference continuing mechanism to the `PrestoDriver` class. It doesn't seem to be the responsibility of the `QueryExecutor` to count self-references. Generally it can be used somewhere else, where reference counting isn't needed.